### PR TITLE
Feature/epmlabsbrn 885 add exercise api

### PR DESCRIPTION
**Pull request by Dmitriy1Fokin**
[EPMLABSBRN-885](https://jira.epam.com/jira/browse/EPMLABSBRN-885)

**Description**:
design & implement rest api for adding new exercise for exist subgroup

### CODE REVIEW

File: api-contract/api.raml [hunk 1]
```diff
@@ -819,6 +819,39 @@ mediaType:
                 body:
                   application/json:
                     example: !include samples/add_subgroup_to_series_response.json
+    /create/exercise/words:
```
Comment (ElenaSpb) [inline api-contract/api.raml:822]: maybe it is better to get type as a parameter?
so on new type we will need not to add additional end-point. ?

File: src/main/kotlin/com/epam/brn/upload/csv/seriesPhrases/SeriesPhrasesRecordProcessor.kt [hunk 2]
```diff
@@ -37,28 +37,24 @@ class SeriesPhrasesRecordProcessor(
...
+                val subGroup = subGroupRepository.findByCodeAndLocale(it.code, locale.locale)
+                    ?: throw EntityNotFoundException("No subGroup was found for code=${it.code} and locale={${locale.locale}}")
 
-        records.forEach { record ->
-            val subGroup = subGroupRepository.findByCodeAndLocale(record.code, locale.locale)
-                ?: throw EntityNotFoundException("No subGroup was found for code=${record.code} and locale={${locale.locale}}")
-            val existExercise = exerciseRepository.findExerciseByNameAndLevel(record.exerciseName, record.level)
-            if (!existExercise.isPresent) {
-                val answerOptions = extractAnswerOptions(record, locale)
+                val answerOptions = extractAnswerOptions(it, locale)
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/upload/csv/seriesPhrases/SeriesPhrasesRecordProcessor.kt:48]: why this refactoring was done in this task which is not related to processors. ?
also logic was changed.. 
I think all changes about loading data/exercises should be done in separate task and in separate MR if changes would be neadable/reasonable. 
I did not understand what was the reason of this changes..

File: src/main/kotlin/com/epam/brn/dto/request/exercise/ExerciseType.kt [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+enum class ExerciseType {
+    WORDS_SEQUENCES,
+    SENTENCE,
+    SINGLE_SIMPLE_WORDS,
+    PHRASES,
+    FREQUENCY_WORDS,
+    DI,
+    DURATION_SIGNALS,
+    FREQUENCY_SIGNALS
+}
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/dto/request/exercise/ExerciseType.kt:12]: we already have this enum in package com.epam.brn.model
why do you add the same  one?

File: src/main/kotlin/com/epam/brn/service/ExerciseService.kt [hunk 4]
```diff
@@ -131,4 +145,60 @@ class ExerciseService(
         val subGroupExercises = exerciseRepository.findExercisesBySubGroupId(subGroupId)
         return subGroupExercises.map { it.toDtoWithTasks() }
     }
+
+    @Transactional(rollbackFor = [Exception::class])
+    fun createAndGenerateExercise(exerciseCreateDto: ExerciseCreateDto): ExerciseDto {
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/service/ExerciseService.kt:150]: 'createAndGenerate'
too same words )

File: src/main/kotlin/com/epam/brn/dto/request/exercise/ExerciseType.kt (hunk 3 above)
Comment (Dmitriy1Fokin) [reply src/main/kotlin/com/epam/brn/dto/request/exercise/ExerciseType.kt:12]: I didn’t know that there was such enum. Deleted this

File: src/main/kotlin/com/epam/brn/service/ExerciseService.kt (hunk 4 above)
Comment (Dmitriy1Fokin) [reply src/main/kotlin/com/epam/brn/service/ExerciseService.kt:150]: Ok. I don't mind

File: src/main/kotlin/com/epam/brn/validator/WithoutPunctuationValidator.kt [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package com.epam.brn.validator
+
+import javax.validation.ConstraintValidator
+import javax.validation.ConstraintValidatorContext
+import kotlin.text.Regex
+
+class WithoutPunctuationValidator : ConstraintValidator<WithoutPunctuation, String> {
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/validator/WithoutPunctuationValidator.kt:7]: why WithoutPunctuationValidator
if it is PunctuationValidator
?

File: src/main/kotlin/com/epam/brn/dto/request/exercise/ExerciseCreateDto.kt [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    @ApiModelProperty(value = "Exercise name", required = true, example = "Фразы разной длительности")
+    @field:NotBlank
+    open val exerciseName: String
+)
+
+@ApiModel(value = "ExerciseWordsCreateDto", description = "Request dto for create exercise 'words'")
+data class ExerciseWordsCreateDto(
+    @ApiModelProperty(value = "Locale", required = true, example = "RU")
+    @field:NotNull
+    override val locale: Locale,
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/dto/request/exercise/ExerciseCreateDto.kt:49]: is it really needable to add parent fields into childs?

File: src/main/kotlin/com/epam/brn/service/ExerciseService.kt [hunk 7]
```diff
@@ -131,4 +145,60 @@ class ExerciseService(
...
         return subGroupExercises.map { it.toDtoWithTasks() }
     }
+
+    @Transactional(rollbackFor = [Exception::class])
+    fun createExercise(exerciseCreateDto: ExerciseCreateDto): ExerciseDto {
+        val exercise = when (exerciseCreateDto) {
+            is ExerciseWordsCreateDto -> {
+                val seriesWordsRecord = exerciseCreateDto.toSeriesWordsRecord()
+                val exercise = createExercise(seriesWordsRecord, exerciseCreateDto.locale)
+                    ?: throw RuntimeException("Exercise with this name (${exerciseCreateDto.exerciseName}) already exist")
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/service/ExerciseService.kt:155]: maybe better IllegalArgumetException by semantic fit.?

File: src/main/kotlin/com/epam/brn/validator/WithoutPunctuationValidator.kt (hunk 5 above)
Comment (Dmitriy1Fokin) [reply src/main/kotlin/com/epam/brn/validator/WithoutPunctuationValidator.kt:7]: fixed

File: src/main/kotlin/com/epam/brn/dto/request/exercise/ExerciseCreateDto.kt (hunk 6 above)
Comment (Dmitriy1Fokin) [reply src/main/kotlin/com/epam/brn/dto/request/exercise/ExerciseCreateDto.kt:49]: Unfortunately, yes. [Proof](https://discuss.kotlinlang.org/t/sealed-class-inheritance-and-constructors/2283)

File: src/main/kotlin/com/epam/brn/service/ExerciseService.kt (hunk 7 above)
Comment (Dmitriy1Fokin) [reply src/main/kotlin/com/epam/brn/service/ExerciseService.kt:155]: fixed

**Review by ElenaSpb — APPROVED** [review]
